### PR TITLE
Nare/c++/19 mar2021

### PR DESCRIPTION
**Pull request by nare99**
Added homework without templates

### CODE REVIEW

File: Nare/19_03/complex.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                return comp << a.real << "+" << a.image << "i";
+            }
+            else
+                return comp << a.real << "" << a.image << "i";
+            }
+        else {
+            return comp << a.real;
+        }
+    }
+    else {
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: Delete extra 'else'

File: Nare/19_03/complex.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Complex {
+    private:
+        int real;
+        int image;
```
Comment (mKolibri) [inline Nare/19_03/complex.h]: Please implement setters and getters :)

File: Nare/19_03/main.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    /*
+     Constructor with parametrs
+    */ 
+    Complex a(5,2);
+    Complex b(3,4);
+
+    /*
+     Constructor without parametrs
+    */ 
+    Complex a1, b1;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Declare every new variable in new line

File: Nare/19_03/main.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    
+    /*
+     Use of overloaded operator>>
+    */ 
+    std::cout << "Use of overloaded operator>>" << std::endl << "(Enter 2 complex numbers)" << std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note ** Use 2 cout for info, this one is so long :)

File: Nare/19_03/main.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    std::cin >> a1;
+    std::cin >> b1;
+    std::cout << std::endl;
+    /*
+     Copy constructor and overloaded operator<<
+    */
+    std::cout << "Use of copy constructor(copies value of b` 3+4i to c) and overloaded operator<<"<<std::endl;
+    Complex c = b;
+    std::cout << c <<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: std::cout << c << std::endl;
You lost 'space'

File: Nare/19_03/main.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << std::endl;
+    /*
+     Use of overloaded operator+
+    */ 
+    std::cout << "Count (5+2i)+(3+4i)"<<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note * bad codding style for this line

File: Nare/19_03/main.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << a + b << std::endl;
+    std::cout << std::endl;
+
+    /*
+     Use of overloaded operator-
+    */ 
+    std::cout << "Count (5+2i)-(3+4i)"<<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note *

File: Nare/19_03/main.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << a - b << std::endl;
+    std::cout << std::endl;
+
+    /*
+     Use of overloaded operator*
+    */ 
+    std::cout << "Count (5+2i)*(3+4i)"<<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note *

File: Nare/19_03/main.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << a * b << std::endl;
+    std::cout << std::endl;
+
+    /*
+     Use of overloaded operator/
+    */ 
+    std::cout << "Count entered complex numbers` (" << a1 << ")/(" << b1 << ") division(returns int)" << std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note **

File: Nare/19_03/main.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << a1 / b1 << std::endl;
+    std::cout << std::endl;
+
+    /*
+     Use of function with static variable
+    */ 
+    std::cout << "Called print function with static variable 4 times"<<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note *

File: Nare/19_03/main.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    print();
+    print();
+    print();
+    std::cout<<print()<<std::endl;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Note *

File: Nare/19_03/main.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+return 0;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: there should be 4 spaces at the beginning of line

File: Nare/19_03/complex.cpp [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    Complex c;
+    c.real = (this -> real * a.real)  - (this -> image * a.image);
+    c.image = (this -> real * a.image) + (this -> image * a.real);
+    return c;
+}
+
+/*
+ Overload operator/
+*/ 
+Complex Complex::operator/(const Complex& a) {
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: Good work :100:

File: Nare/19_03/complex.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    Complex c;
+
+    c.real = ((this -> real * a.real)  + (this -> image * a.image))/((a.real * a.real) + (a.image * a.image));
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: use variables for this lines, it is so long

File: Nare/19_03/complex.cpp [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+    else {
+        return comp << a.image << "i";
+    }
+
+}
+
+/*
+ Overload operator<<
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: wrong comment :)

File: Nare/19_03/complex.cpp [hunk 16]
```diff
@@ -0,0 +1,97 @@
+#include <iostream>
+#include "complex.h"
+
+/*
+  Constructor witout paramerts
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp:5]: Do you mean Default constructor ? :)

File: Nare/19_03/complex.cpp [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    Complex c;
+    c.real = this -> real - a.real;
+    c.image = this -> image - a.image;
+    return c;
+}
+
+/*
+ Overload operator*
+*/ 
+
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: extra line

File: Nare/19_03/complex.h [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        Complex operator/(const Complex& a);
+
+        /*
+         Overload <<, >> operators
+        */ 
+        friend std::ostream& operator<<(std::ostream& comp, const Complex& a);
+        friend std::istream& operator>>(std::istream& comp, Complex& a);
+
+        /*
+         function with static member
```
Comment (mKolibri) [inline Nare/19_03/complex.h]: Extra comment ? :)

File: Nare/19_03/complex.h [hunk 19]
```diff
@@ -0,0 +1,31 @@
+class Complex {
```
Comment (mKolibri) [inline Nare/19_03/complex.h:9]: please always write header files with #ifndef, .... #endif

File: Nare/19_03/main.cpp [hunk 20]
```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include "complex.h"
+
+int print();
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: write on header file

File: Nare/19_03/main.cpp [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    print();
+    print();
+    std::cout<<print()<<std::endl;
+return 0;
+}
+
+/*
+  Function with static variable 
+*/
+ int print() {
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: write in other cpp file !

File: Nare/19_03/main.cpp [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+     static int count = 0;
+     ++count;
```
Comment (mKolibri) [inline Nare/19_03/main.cpp]: Instead of { ++count; return count; }
you can use -> { return count++; }

File: Nare/19_03/complex.cpp [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    c.real = ((this -> real * a.real)  + (this -> image * a.image))/((a.real * a.real) + (a.image * a.image));
+    c.image = ((this -> real * (-a.image))  + (this -> image * a.real))/((a.real * a.real) + (a.image * a.image));
+    return c;
+}
+
+/*
+ Overload operator<<
+*/ 
+std::ostream& operator<<(std::ostream& comp, const Complex& a) {
+    if(a.real != 0) {
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: after if must be state.

File: Nare/19_03/complex.cpp [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+        if(a.image != 0 ) {
+            if(a.image > 0) {
+                return comp << a.real << "+" << a.image << "i";
+            }
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: Please optimize this function

**Review by mKolibri — CHANGES_REQUESTED** [review]
Your codding style not good, but you can be more carefully.
Also please optimize the code, and you need to remember that Git can not diff binary file. It will upload entry file into repository and will store it's pretty much forever. Delete it.

File: Nare/19_03/complex.cpp [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+/*
+ Overload operator/
+*/ 
+Complex Complex::operator/(const Complex& a) {
+    Complex c;
+
+    c.real = ((this -> real * a.real)  + (this -> image * a.image))/((a.real * a.real) + (a.image * a.image));
+    c.image = ((this -> real * (-a.image))  + (this -> image * a.real))/((a.real * a.real) + (a.image * a.image));
```
Comment (mKolibri) [inline Nare/19_03/complex.cpp]: you must check if second part is not 0, or you can have division by 0 exception.

**Review by mKolibri — CHANGES_REQUESTED** [review]
Your codding style not good, but you can be more carefully.
Also please optimize the code, and you need to remember that Git can not diff binary file. It will upload entry file into repository and will store it's pretty much forever. Delete it.

File: Nare/19_03/main.cpp (hunk 9 above)
Comment (nare99) [reply Nare/19_03/main.cpp]: Can i use 3 cout?`
std::cout << "Count entered complex numbers`"; 
std::cout << "(" << a1 << ")/(" << b1 << ")"; 
std::cout << "division(returns int)" << std::endl;

File: Nare/19_03/complex.cpp (hunk 16 above)
Comment (nare99) [reply Nare/19_03/complex.cpp:5]: Yes.

File: Nare/19_03/complex.h (hunk 18 above)
Comment (nare99) [reply Nare/19_03/complex.h]: Print function first was written as class method, then I forget to delete comment.

File: Nare/19_03/main.cpp (hunk 22 above)
Comment (nare99) [reply Nare/19_03/main.cpp]: return count++ returns count's previous value. I will write ++count to get real count of function calls

**Pull request merged**